teps/0065-retry-failed-tasks-on-demand.md

```diff
@@ -0,0 +1,238 @@
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+    - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Notes/Caveats (optional)](#notescaveats-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+    - [User Experience (optional)](#user-experience-optional)
+    - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Presently, a pipeline has a mechanism for `retry` a task, which a pipeline author
+can configure at the time of creation of a `Pipeline` or a `PipelineRun`. In this
+TEP, we are exploring the benefits of adding a new API `retry` which will allow a
+user to "on-demand", retry a failed pipeline run. In other words, 
+`rerun failed tests` from CI/CD world.
+
+## Motivation
+
+**Optimise the use of cluster resources.**
+
+_Why do we need a new `retry` API when we already support retry in `Pipeline`
+tasks?_ 
+The present retry field can only be defined at the time of creation of pipeline
+and not, as an `on-demand` invocation? This is not suitable for use cases, where
+a manual intervention is necessary to decide whether a rerun is required. For 
+example a pull request, may have some test suit failures or stress test failures
+with known flakes in recent times. Now, we need a way to know whether a failure
+was due to some regression in the patch that pull-request proposes, or it is due
+to flakiness of jobs itself. In this case, simply retrying `n` number of times
+does not seem to help with optimal resource consumption.
+
+For example, At present `/retest` at kubernetes/kubernetes repo reruns only
+the failed jobs, a new api for retrying failed `pipelineRun` will give out
+of the box support. Hope they use `tektoncd` as backend for their CI at some
+point.
+
+Without this support, at present a pull-request author or reviewer has to
+individually, mark tests for rerun.
+
+### Goals
+
+1. Explore both the merits and demerits in having a new API for on-demand
+   retrying, an _only a failed_ pipeline.
+2. A pipeline may either have failed due to some failures in the tasks or may
+   be user invoked cancel request. Retry only the failed/canceled tasks for a
+   failed `pipelineRun`.
+3. Document the feature in the tekton documentation, explaining the use cases.
+
+### Non-Goals
+
+1. Retry of successful pipeline runs or anything other than a failed pipeline
+   run.
+2. Changing or discussing such a possibility of existing retry mechanism.
+3. Discuss checkpointing of pipeline state or workspaces, etc. A `pipelineRun`'s
+   state stored in etcd is used as is.
```

Comment:
For this feature to work, the pipeline and taskruns in the retry execution will need access to the state of the previous run, so in case any of the state is deleted, the re-run will fail:
- writable workspaces needs to be still available:
   - those provisioned by tekton will stick around until the pipelinerun is not deleted
   - those provided by users are not controller by tekton
- results will be store in etcd in the original pipelinerun

Even workspaces + results might not be enough though, since the initial tasks or a pipeline might setup state outside of the workspace (e.g. create a test cluster, provision a test service), and that state will be usually cleaned-up by the `finally` part of the pipeline. If the URI of the resources provisioned by initial tasks is exposed as results, those URIs can be passed to the new execution and if still valid the new execution could succeed.